### ISSUE TITLE
AMP "LiveFire" unit test setup contains race condition

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1965]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1965|
|Type|defect|
|Created|2006-07-28 02:27:45Z|

```LiveFireBase``` has a ```setUp``` that looks like this:

```
    def setUp(self):
        from twisted.internet import reactor
        self.serverFactory = protocol.ServerFactory()
        self.serverFactory.protocol = self.serverProto
        self.clientFactory = protocol.ClientFactory()
        self.clientFactory.protocol = self.clientProto
        self.clientFactory.onMade = defer.Deferred()
        self.serverPort = reactor.listenTCP(0, self.serverFactory)
        self.clientConn = reactor.connectTCP(
            '127.0.0.1', self.serverPort.getHost().port,
            self.clientFactory)
        def getProtos(rlst):
            self.cli = self.clientFactory.theProto
            self.svr = self.serverFactory.theProto
        return self.clientFactory.onMade.addCallback(getProtos)
```

This is incorrect: it assumes that the server will have had buildProtocol called by the time the client's connectionMade callback runs.  The code needs to wait on a Deferred for each side of the connection.

I haven't looked at the rest of the tests so I don't know of similar races are present elsewhere.

Technically this is a test suite regression and the AMP branch should be reverted, but the problem appears to be purely in the tests and the failure is only triggered on one reactor.  Assuming this is fixed ASAP it's probably okay not to revert the branch.

Someone should go over the rest of the tests AMP added, though, and make sure there are no other similar problems.

<details><summary>Searchable metadata</summary>

```
trac-id__1965 1965
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1154053665000000 1154053665000000
changetime__1154365981000000 1154365981000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

Starting work on this now in ```patience-1965```.

FWIW, the live fire tests are the only ones that do anything with actual sockets, so unless you find another such race condition in tearDown it is unlikely that the other tests are going to cause problems.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

OK, that branch is ready for review.  Should be a quickie for anyone interested, but assigning to exarkun as the reporter.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @glyph|
|-|-|

The tearDown on the same class also incorrectly assumes stopListening always returns a Deferred.  I think the use of a DeferredList like that is also somewhat sketchy.  It needs to block the ConnectionLost/Done errback, but it is also consuming and discarding all other errors without checking their types.  I doubt this is masking anything or will ever mask anything, but it's somewhat sloppy.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

To clarify, fix the tearDown Deferred handling bug and then merge it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `closed`|
|-|-|

(In [17778]) Fix race conditions and test ugliness in test_amp.

Author: glyph

Reviewer: exarkun

Fixes [#1965](https://github.com/twisted/twisted/issues/1965)

There were 2 bugs in test_amp's setUp and tearDown; this fixes them.  First,
the QT reactor and win32event reactor buildbots were failing due to a race
condition between the client and the server's connectionMade methods, and
second, 'stopListening' doesn't always return a Deferred, so the tests now deal
with that.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
